pkg/solana/txm/pendingtx.go

```diff
@@ -138,28 +196,270 @@ func (c *pendingTxContext) ListAll() []solana.Signature {
+	// upgrade to write lock if sig and id exist
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	id, sigExists = c.sigToID[sig]
+	if !sigExists {
+		return id, errors.New("signature does not exist")
+	}
+	if _, exists := c.broadcastedTxs[id]; !exists {
+		return id, errors.New("id does not exist in brooadcasted map")
+	}
+	tx = c.broadcastedTxs[id]
+	// update tx state to Processed
+	tx.state = Processed
+	// save updated tx back to the broadcasted map
+	c.broadcastedTxs[id] = tx
+	return id, nil
 }
 
-func (c *pendingTxContext) OnSuccess(sig solana.Signature) uuid.UUID {
-	return c.Remove(sig)
+func (c *pendingTxContext) OnConfirmed(sig solana.Signature) (string, error) {
+	c.lock.RLock()
+	// validate if sig exists
+	id, sigExists := c.sigToID[sig]
+	if !sigExists {
+		c.lock.RUnlock()
+		return id, errors.New("signature does not exist")
+	}
+	// Check if transaction already in confirmed state
+	if tx, exists := c.confirmedTxs[id]; exists && tx.state == Confirmed {
+		c.lock.RUnlock()
+		return id, ErrAlreadyInExpectedState
+	}
+	// Transactions should only move to confirmed from broadcasted/processed
+	if _, exists := c.broadcastedTxs[id]; !exists {
+		c.lock.RUnlock()
+		return id, errors.New("id does not exist in broadcasted map")
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock if id exists
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	id, sigExists = c.sigToID[sig]
+	if !sigExists {
+		return id, errors.New("signature does not exist")
+	}
+	if _, exists := c.broadcastedTxs[id]; !exists {
+		return id, errors.New("id does not exist in broadcasted map")
+	}
+	// call cancel func + remove from map to stop the retry/bumping cycle for this transaction
+	if cancel, exists := c.cancelBy[id]; exists {
+		cancel() // cancel context
+		delete(c.cancelBy, id)
+	}
+	tx := c.broadcastedTxs[id]
+	// update tx state to Confirmed
+	tx.state = Confirmed
+	// move tx to confirmed map
+	c.confirmedTxs[id] = tx
+	// remove tx from broadcasted map
+	delete(c.broadcastedTxs, id)
+	return id, nil
 }
 
-func (c *pendingTxContext) OnError(sig solana.Signature, _ int) uuid.UUID {
-	return c.Remove(sig)
+func (c *pendingTxContext) OnFinalized(sig solana.Signature, retentionTimeout time.Duration) (string, error) {
+	// check if already cancelled
+	c.lock.RLock()
+	id, sigExists := c.sigToID[sig]
+	if !sigExists {
+		c.lock.RUnlock()
+		return id, errors.New("signature does not exist")
+	}
+	// Allow transactions to transition from broadcasted, processed, or confirmed state in case there are delays between status checks
+	var broadcastedExists, confirmedExists bool
+	_, broadcastedExists = c.broadcastedTxs[id]
+	_, confirmedExists = c.confirmedTxs[id]
+	if !broadcastedExists && !confirmedExists {
+		c.lock.RUnlock()
+		return id, errors.New("id does not exist in broadcasted or confirmed map")
+	}
+	c.lock.RUnlock()
+
+	// upgrade to write lock if id exists
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if _, exists := c.sigToID[sig]; !exists {
```

Comment:
nit: would make sense to abstract these checks that are used across multiple places?

```
func (c *pendingTxContext) signatureExists(sig solana.Signature) bool {
    _, exists := c.sigToID[sig]
    return exists
}

func (c *pendingTxContext) idExists(id string) bool {
    _, exists := c.broadcastedTxs[id]
    return exists
}
```

and then use them as:
```
if c.signatureExists(sig) {
    return ErrSignatureExists
}
if !c.idExists(tx.id) {
    return ErrIDNotExists
}
```

Reply:
This could help in a few places but in others we also need the `id` from this check too where this wouldn't work. I don't think it's worth abstracting since we just replace the map access with a method call so we aren't really saving any lines either.